src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs

```diff
@@ -400,19 +453,65 @@ public override BoundNode VisitBlock(BoundBlock node)
             var parentBlock = _currentBlock;
             _currentBlock = node;
             var initialUsingCount = _usingDeclarations.Count;
+            var hadUsingDeclarations = false;
             foreach (var local in node.Locals)
             {
                 if (local.IsUsing)
                 {
                     _usingDeclarations.Add((local, node));
+                    hadUsingDeclarations = true;
                 }
             }
 
             var result = base.VisitBlock(node);
 
             _usingDeclarations.Clip(initialUsingCount);
+            if (_asyncTryFinallyEndsReachable is not null && hadUsingDeclarations)
+            {
+                UsingDeclarationVisitor.Visit(node, _asyncTryFinallyEndsReachable, State.Alive);
+            }
             _currentBlock = parentBlock;
             return result;
         }
+
+        private sealed class UsingDeclarationVisitor : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            internal static void Visit(BoundBlock block, PooledDictionary<BoundNode, bool> dictionary, bool alive)
+            {
+                var instance = new UsingDeclarationVisitor(dictionary, alive);
+                instance.VisitList(block.Statements);
+            }
+
+            private UsingDeclarationVisitor(PooledDictionary<BoundNode, bool> dictionary, bool alive)
+            {
+                _dictionary = dictionary;
+                _alive = alive;
+            }
+
+            private readonly PooledDictionary<BoundNode, bool> _dictionary;
+            private readonly bool _alive;
+
+#pragma warning disable CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring as nullable.
+            private UsingDeclarationVisitor()
+#pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring as nullable.
+            {
+            }
+
+            public override BoundNode? VisitUsingLocalDeclarations(BoundUsingLocalDeclarations node)
+            {
+                if (node.AwaitOpt is not null)
+                {
+                    _dictionary.Add(node, _alive);
+                }
+
+                return base.VisitUsingLocalDeclarations(node);
+            }
+
+            public override BoundNode? VisitBlock(BoundBlock node)
+            {
+                // Do not recurse through nested blocks, they will be visited by their own visits through the ControlFlowPass.
```

Comment:
It may be good to mention and even test the scenario with embedded statement (a using declaration within a statement but without a block). Do we reach the `UsingDeclarationVisitor` in such error case?